RecoEcal/EgammaClusterAlgos/src/SCEnergyCorrectorDRN.cc

```diff
@@ -0,0 +1,116 @@
+  setTokens(iConfig, cc);
+}
+
+void SCEnergyCorrectorDRN::fillPSetDescription(edm::ParameterSetDescription& desc) {
+  desc.add<edm::InputTag>("ecalRecHitsEE", edm::InputTag("ecalRecHit", "EcalRecHitsEE"));
+  desc.add<edm::InputTag>("ecalRecHitsEB", edm::InputTag("ecalRecHit", "EcalRecHitsEB"));
+  desc.add<edm::InputTag>("rhoFastJet");
+}
+
+edm::ParameterSetDescription SCEnergyCorrectorDRN::makePSetDescription() {
+  edm::ParameterSetDescription desc;
+  fillPSetDescription(desc);
+  return desc;
+}
+
+void SCEnergyCorrectorDRN::setEventSetup(const edm::EventSetup& es) {
+  caloTopo_ = &es.getData(caloTopoToken_);
+  caloGeom_ = &es.getData(caloGeomToken_);
+}
+
+void SCEnergyCorrectorDRN::setEvent(const edm::Event& event) {
+  event.getByToken(tokenEBRecHits_, recHitsEB_);
+  event.getByToken(tokenEERecHits_, recHitsEE_);
+  event.getByToken(rhoToken_, rhoHandle_);
+}
+
+void SCEnergyCorrectorDRN::makeInput(const edm::Event& iEvent, TritonInputMap& iInput, const reco::SuperClusterCollection& inputSCs ) const {
+
+    std::vector<unsigned> nHits = {};
+    unsigned totalHits = 0;
+    unsigned n;
+    for(const auto& inputSC : inputSCs){
+        n = inputSC.hitsAndFractions().size();
+        totalHits+=n;
+        nHits.push_back(n);
+    }
+
+    //set shapes
+    auto& input1 = iInput.at("x__0");
+    input1.setShape(0, totalHits);
+    auto data1 = input1.allocate<float>();
+    auto& vdata1 = (*data1)[0];
+
+    auto& input2 = iInput.at("batch__1");
+    input2.setShape(0, totalHits);
+    auto data2 = input2.allocate<int64_t>();
+    auto& vdata2 = (*data2)[0];
+
+    auto& input3 = iInput.at("graphx__2");
+    input3.setShape(0, 2*nHits.size());
+    auto data3 = input3.allocate<float>();
+    auto& vdata3 = (*data3)[0];
+
+    //fill
+    unsigned batchNum = 0;
+    float En, frac, x, y, z;
+    for(const auto& inputSC : inputSCs){
+        const auto& hits = inputSC.hitsAndFractions();
+        const bool isEB = hits[0].first.subdetId() == EcalBarrel;
+        const auto& recHitsProduct = isEB ? recHitsEB_.product() : recHitsEE_.product();
+        for(const auto& hit : hits){
+            En = EcalClusterTools::recHitEnergy(hit.first, recHitsProduct);
+            frac = hit.second;
+            GlobalPoint position = caloGeom_->getGeometry(hit.first)->getPosition();
+            x = (position.x()+150.0)/300.0;
+            y = (position.y()+150.0)/300.0;
+            z = (position.z()+330.0)/660.0;
+            vdata1.push_back(x);
+            vdata1.push_back(y);
+            vdata1.push_back(z);
+            vdata1.push_back(En*frac/250.0);
+            vdata2.push_back(batchNum);
+        }
+        vdata3.push_back(*rhoHandle_/ RHO_MAX );
+        vdata3.push_back(0.0);
+        ++batchNum;
+    }
+
+    // convert to server format
+    input1.toServer(data1);
+    input2.toServer(data2);
+    input3.toServer(data3);
+}
+
+TritonOutput<float> SCEnergyCorrectorDRN::getOutput(const TritonOutputMap& iOutput) {
```

Comment:
I tend to agree here. It looks like SCEnergyCorrectorDRN is Triton-specific, so it would make more sense to put the evaluation code directly in SCEnergyCorrectorDRNProducer.

On the other hand, if the idea was to separate the NN evaluation boilerplate from the actual inference framework (to e.g. support different backends), the Triton-specific parts should be removed from SCEnergyCorrectorDRN.

Reply:
If it is okay, I think I would like to keep this helper class as-is. It is unfortunately not possible to decouple the Triton-specific code from the producer itself, but keeping the ML input/output in the helper class allows the producer to be more agnostic to the details of the architecture being used for the regression (which may be subject to further development in the future).

Reply:
Having a Triton producer that can handle different, related architectures is reasonable, but a lot of the choices in this particular design do not seem related to that goal. Do you expect to change the actual CMSSW input collections used?

Reply:
For example, we are currently considering adding in information from the RecHit flags, gain channel used, etc. This would require a meaningful update to the data passed to the network, and to the format of the inputs expected by the model.

Reply:
Okay, if the actual CMS inputs may change, then the delegation of consumes calls etc. to the helper makes sense.